crates/chia-datalayer/src/merkle/iterators.rs

```diff
@@ -61,13 +63,39 @@ impl Iterator for LeftChildFirstIterator<'_> {
                 }
             }
 
+            match block.node.parent().0 {
+                // TODO: maybe also check the parent considers this the child?
+                Some(index) => {
+                    if item.index != self.from_index && !self.already_queued.contains(&index) {
+                        return Some(Err(Error::ReferenceToUnknownParent()));
+                    }
+                }
+                None => {
+                    if item.index.0 != 0 {
+                        return Some(Err(Error::UnexpectedParentlessNode()));
+                    }
+                }
+            }
+
             match block.node {
-                Node::Leaf(..) => return Some(Ok((item.index, block))),
+                Node::Leaf(..) => {
+                    if block.metadata.dirty {
```

Comment:
I got the impression that this would be calculated on-demand. It might be tricky to do while iterating over the tree immutably. There are `RefCell`s though.